### ISSUE TITLE
packFolders configuration is being cached

### ISSUE THREAD

**Issue by Pjb518** [foundryvtt/foundryvtt#9309]
### What happened?

While testing the new `packFolders` configuration for the `a5e` system, I noticed that changes to the configuration are only reflected in worlds that have not already been loaded with `packFolders` config.

The following three images show three worlds which are all using the same system, with the same `system.json`.

<img width="306" alt="Screenshot 2023-04-28 at 16 59 24" src="https://user-images.githubusercontent.com/33249712/235197876-c2822f20-f7d3-44dc-9d0f-bf6758d6519e.png">
<img width="306" alt="Screenshot 2023-04-28 at 16 58 40" src="https://user-images.githubusercontent.com/33249712/235197891-e0667236-9bcc-4a22-ba35-cc937d3201d0.png">
<img width="307" alt="Screenshot 2023-04-28 at 16 57 52" src="https://user-images.githubusercontent.com/33249712/235197911-e1713ee1-f771-470b-9742-c2b6645aaf3d.png">

The current configuration looks like this:
```json
"packFolders": [
  {
    "name": "A5E Core",
    "sorting": "m",
    "packs": [
      "a5e-monsters",
      "a5e-adventuring-gear",
      "a5e-magic-items",
      "a5e-maneuvers",
      "a5e-spells",
      "a5e-rare-spells"
    ],
    "folders": [
      {
        "name": "Player Options",
        "sorting": "m",
        "packs": [
          "a5e-backgrounds",
          "a5e-cultures",
          "a5e-destinies",
          "a5e-feats",
          "a5e-synergy-feats",
          "a5e-background-features",
          "a5e-class-features",
          "a5e-culture-features",
          "a5e-destiny-features",
          "a5e-heritage-features",
          "a5e-paragon-gifts"
        ]
      }
    ]
  }
]
```

Despite all three worlds running the same system, with the same `packFolders` configuration defined in the `system.json`, the worlds in the first two images still reflect older versions of my config.

Incrementing the system version doesn't cause the folder structure to update. Even if it did, this would be problematic for testing, as every change would require a new system or module version number.

The expected behaviour is that the compendium folder structure should reflect the current version of the `packFolders` config in the `module.json` or `system.json`.

### What ways of accessing Foundry can you encounter this issue in?

- [X] Native App (Electron)
- [X] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Other

### Reproduction Steps

- Use a module or system with compendium packs.
- Set some configuration for `packFolders` in the `module.json` or `system.json`.
- Open a world running this module or system and confirm the compendium config is working.
- Close Foundry.
- Change the configuration of `packFolders` in the `module.json` or `system.json`.
- Reopen the same world and see if the compendia reflect the latest config.

### What core version are you reporting this for?

Version 11 Testing 1 (Build 296)

### Bug Checklist

- [X] The issue occurs while all Modules are disabled

**Comment by stwlam** [comment]
Perhaps manifest-defined folders should be in-memory-only and read-only in a world. That would permit a system or module to reorganize them freely and have it such changes reflected in all worlds upon package update.

**Comment by aaclayton** [comment]
We need to do a bit more thinking about this workflow and decide whether we want to treat this as a one-time only workflow or an assignment that can evolve as modules and systems do.

**Comment by Pjb518** [comment]
> I'm not sure there's anything I want to change here - for developing, you can programmatically clear your foldering to apply the new default, but outside of that, it's probably best that a GM has final say over their foldering setup, and that their compendiums don't refolder on them when they update a package.

I would strongly prefer for the packs to respect the configuration in the manifest except in cases where the user has manually altered the folder structure.

You're right that we may be able to programmatically clear any foldering configuration in development (although this is not an obvious or particularly nice workflow), but that doesn't really help if we want to change the default structure between releases. In the A5e system, the content we work with is very permissively licensed, and our compendia are going to grow substantially over coming year. The large number of additions may require a shift in our organisational structure, to ensure the large number of compendia remain navigable. Ideally, we'd also like to test certain structures with our users and modify them according to user feedback. We can't really do that right now, because any changes are not going to be reflected for our existing users.

I'm also not really sure what the intended behaviour is when it comes to adding new packs to an existing `packFolders` config, or adding new folders that only include new packs. What we if add a new folder that only includes old packs? Does that create an empty folder? I can also see a lot of mess being created by new packs that are nested inside subfolders in worlds where users have modified the folder structure.

Thanks for taking a look at this.

**Comment by cswendrowski** [comment]
> I would strongly prefer for the packs to respect the configuration in the manifest except in cases where the user has manually altered the folder structure.

We mostly agree here, excepting that - as of right now - Foundry only knows if a Compendium is foldered or not, not if it's foldered in a previously defined `packFolders` that no longer exists in this version. We could track additional flag metadata to say "this has not been modified by the user yet", except we currently don't keep track of what operations are API-driven versus User-driven, so we would have to add that concept, which is complicated to say the least. Even if we had that data, I still personally think it's confusing as a GM to update a module and have my compendium packs reorganized on me after I've already gotten used to its existing location. Users can reorganize to match updated structures if they choose to, even via macro, but I don't think we should force reorganization upon users

> What if we add a new folder that only includes old packs? Does that create an empty folder?

New folders are only created when a Compendium would be nested under them. If a new folder would seat an old pack that's already foldered, that new folder wouldn't be created for that user.

**Comment by Pjb518** [comment]
> Even if we had that data, I still personally think it's confusing as a GM to update a module and have my compendium packs reorganized on me after I've already gotten used to its existing location.

I disagree. There are already plenty of ways that the packs could be reorganised without considering `packFolders`, and that doesn't seem to be a concern. Packs could be renamed, for example, or packs could be split into multiple packs / merged into one. The latter kind of change is planned to take place in the A5e system to better indicate which content the material comes from. We just couldn't justify this without having the ability to organise things into folders, as we already flood the compendium tab as it is.

There's also the fact that the initial organisation into folders is likely to be just as confusing and disruptive. Even more so if the user is one of the very large number already using the Compendium Folders module to organise their content.

> Users can reorganize to match updated structures if they choose to, even via macro, but I don't think we should force reorganization upon users.

While technically true, how would they know to do this? There's no indication that anything has changed for the end user if a system or module changes its `packFolders` configuration.

A compromise could be to test the current folder structure against the one defined in `packFolders` and to present the user with a dialog allowing them to automatically update to the latest config. The world could also perhaps store the last config the user rejected so that they're not pestered every single upload when nothing has changed. I doubt many modules are regularly changing their `packFolders` config, so this shouldn't cause the user to be bombarded with dialogs.

I worry that not giving the modules and systems any ability to update their default compendium structure for the vast majority of their users is going to lead to a lot of bad practices. As you say, it's possible to force the new config with a macro, so there's nothing stopping modules and systems just hooking into the Foundry setup hooks and nuking the user's folder structure for the module of system's compendium packs. I would very much like to see a more nuanced solution provided by the core software so that quick and dirty measures like that have no reason to exist.

**Comment by cswendrowski** [comment]
After an internal chat, we agree that this could be made better by tracking which folders are assigned via the default and which are manually set by user, then refoldering folders that were in a previous "Default" location. This is a significant overhaul to accomplish, and so won't be part of V11, but we hope to circle back to this in the future

**Comment by elespike** [comment]
Hello! How can I programmatically clear the current foldering in order to apply the new default?

**Comment by Fyorl** [comment]
> Hello! How can I programmatically clear the current foldering in order to apply the new default?

The following will unfolder and re-lock every compendium in the world, which may be all you need.
```js
game.settings.set("core", "compendiumConfiguration", {});
```

If you need a more targeted approach you would need something like this, which will only unfolder all compendia under the given scope:
```js
const config = game.settings.get("core", "compendiumConfiguration");
for ( key in config ) {
  if ( key.startsWith(scope) ) delete config[key].folder;
}
game.settings.set("core", "compendiumConfiguration", config);
```

Then shut down the World and re-launch it.

**Comment by elespike** [comment]
Thanks very much, @Fyorl!
